provisioning/transport/amqp/src/AmqpTransportHandler.cs

```diff
@@ -0,0 +1,145 @@
+            amqpProvider.Versions.Add(AmqpConstants.DefaultProtocolVersion);
+            settings.TransportProviders.Add(amqpProvider);
+            return settings;
+        }
+
+        internal static async Task<AmqpClientConnection> CreateAmqpCloudConnectionAsync(
+            string deviceEndpoint, 
+            X509Certificate2 clientCert, 
+            string linkEndpoint, 
+            bool useWebSocket,
+            ProvisioningSecurityClientSasToken securityClient)
+        {
+            AmqpClientConnection amqpCloudConnection;
+            if (useWebSocket)
+            {
+                // TODO: enable WS
+                amqpCloudConnection = await CreateConnection(
+                    new Uri(WebSocketConstants.Scheme + deviceEndpoint + ":" + WebSocketConstants.Port),
+                    securityClient,
+                    linkEndpoint).ConfigureAwait(false);
+        }
+        else
+            {
+                amqpCloudConnection = await CreateConnection(
+                    new Uri("amqps://" + deviceEndpoint + ":" + AmqpConstants.DefaultSecurePort), 
+                    securityClient, 
+                    linkEndpoint).ConfigureAwait(false);
+            }
+
+            await amqpCloudConnection.OpenAsync(TimeSpan.FromMinutes(1), useWebSocket, clientCert).ConfigureAwait(false);
+
+            var amqpDeviceSession = amqpCloudConnection.CreateSession();
+            await amqpDeviceSession.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            var amqpReceivingLink = amqpDeviceSession.CreateReceivingLink(linkEndpoint);
+
+            amqpReceivingLink.AddClientVersion(clientVersion);
+            amqpReceivingLink.AddApiVersion(apiVersion);
+
+            await amqpReceivingLink.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            var amqpSendingLink = amqpDeviceSession.CreateSendingLink(linkEndpoint);
+
+            amqpSendingLink.AddClientVersion(clientVersion);
+            amqpSendingLink.AddApiVersion(apiVersion);
+
+            await amqpSendingLink.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            return amqpCloudConnection;
+        }
+
+        public static async Task<RegistrationOperationStatus> RegisterDeviceAsync(AmqpClientConnection client, string correlationId)
+        {
+            var amqpMessage = AmqpMessage.Create(new MemoryStream(Encoding.ASCII.GetBytes(DeviceOperations.Register)), false);
+            amqpMessage.Properties.CorrelationId = correlationId;
+            amqpMessage.ApplicationProperties.Map[MessageApplicationPropertyNames.OperationType] = DeviceOperations.Register;
+            amqpMessage.ApplicationProperties.Map[MessageApplicationPropertyNames.ForceRegistration] = false;
+            var outcome = await client.AmqpSession.SendingLink.SendMessageAsync(amqpMessage, new ArraySegment<byte>(Guid.NewGuid().ToByteArray()), TimeSpan.FromMinutes(1)).ConfigureAwait(false);
```

Comment:
From the AMQP source code this function requires 2 ArraySegment structs: deliveryTag and txnId. Are we using an older version?

Reply:
Refer to AmqpClientLink.SendMessageAsyn().  We use AmqpConstants.NullBinary as the txnId.

---
In reply to: [146720881](https://github.com/Azure/azure-iot-sdk-csharp/pull/204#discussion_r146720881) [](ancestors = 146720881)